browser-esbuild/application builds should not show summary stats when the build fails from budget errors

### Command

build

### Is this a regression?

- [X] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

15

### Description

When running "build" command with budget errors, eg `anyComponentStyle`, we expect to be informed in a clear way.

We do see the output of "Application bundle generation failed" in a normal console log color, but the stats table still shows.

### Minimal Reproduction

1. Have any budget that is not met, for instance setting to 1kb with a large style.
2. Run "ng build"
3. Are quietly notified about the build failure, but have to scroll up past all the site chunks in order to see what went wrong.

### Exception or Error

_No response_

### Your Environment

```text
_                      _                 ____ _     ___ 
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | | 
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | | 
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 17.0.3
Node: 20.10.0
Package Manager: npm 10.2.4
OS: win32 x64

Angular: 17.0.4
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router     

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1700.3
@angular-devkit/build-angular   17.0.3
@angular-devkit/core            17.0.3
@angular-devkit/schematics      17.0.3
@angular/cdk                    17.0.1
@angular/cli                    17.0.3
@schematics/angular             17.0.3
rxjs                            7.8.1
typescript                      5.2.2
zone.js                         0.14.2
```


### Anything else relevant?

_No response_

Had some discussion about this in the team today. Our general feeling is that the budget table is actually quite useful to have when a budget is exceeded. This provides a nicely formatted UI to help users understand how a budget was exceeded.

```
$ ng build

⠙ Building...

▲ [WARNING] bundle initial exceeded maximum budget. Budget 1.00 kB was not met by 202.32 kB with a total of 203.32 kB.


✘ [ERROR] bundle initial exceeded maximum budget. Budget 1.00 kB was not met by 202.32 kB with a total of 203.32 kB.

Initial Chunk Files   | Names         |  Raw Size | Estimated Transfer Size
main-45SV4OB3.js      | main          | 170.63 kB |                46.82 kB
polyfills-LZBJRJJE.js | polyfills     |  32.69 kB |                10.59 kB
styles-5INURTSO.css   | styles        |   0 bytes |                 0 bytes

                      | Initial Total | 203.32 kB |                57.41 kB

Application bundle generation failed. [4.568 seconds]
```

In this case we can see that "Initial Total" includes both `main` and `polyfills`, which might not be obvious to users. We could probably do a better job with the general error message, but including the table seems useful in general.

One observation is that it might be better for errors to come *after* the table. Because of the nature of terminal UI autoscroll, most users will read the bottom of the content before the top. The table can also be quite large, so putting errors at the bottom can help avoid users needing to scroll up past the large table before seeing their errors.

@clydin, I think you wanted to look at moving the error messages?

--- a/packages/angular_devkit/build_angular/src/builders/application/execute-build.ts
+++ b/packages/angular_devkit/build_angular/src/builders/application/execute-build.ts
@@ -208,20 +208,12 @@ export async function executeBuild(
   if (options.budgets) {
     const compatStats = generateBudgetStats(metafile, initialFiles);
     budgetFailures = [...checkBudgets(options.budgets, compatStats, true)];
-    if (budgetFailures.length > 0) {
-      const errors = budgetFailures
-        .filter((failure) => failure.severity === 'error')
-        .map(({ message }) => message);
-      const warnings = budgetFailures
-        .filter((failure) => failure.severity !== 'error')
-        .map(({ message }) => message);
-
-      await printWarningsAndErrorsToConsoleAndAddToResult(
-        context,
-        executionResult,
-        warnings,
-        errors,
-      );
+    for (const { message, severity } of budgetFailures) {
+      if (severity === 'error') {
+        executionResult.addError(message);
+      } else {
+        executionResult.addWarning(message);
+      }
     }
   }
 
@@ -234,7 +226,7 @@ export async function executeBuild(
   // Check metafile for CommonJS module usage if optimizing scripts
   if (optimizationOptions.scripts) {
     const messages = checkCommonJSModules(metafile, options.allowedCommonJsDependencies);
-    await logMessages(context, { warnings: messages });
+    executionResult.addWarnings(messages);
   }
 
   // Copy assets
@@ -255,12 +247,10 @@ export async function executeBuild(
 
   // Perform i18n translation inlining if enabled
   let prerenderedRoutes: string[];
-  let errors: string[];
-  let warnings: string[];
   if (i18nOptions.shouldInline) {
     const result = await inlineI18n(options, executionResult, initialFiles);
-    errors = result.errors;
-    warnings = result.warnings;
+    executionResult.addErrors(result.errors);
+    executionResult.addWarnings(result.warnings);
     prerenderedRoutes = result.prerenderedRoutes;
   } else {
     const result = await executePostBundleSteps(
@@ -272,14 +262,12 @@ export async function executeBuild(
       i18nOptions.hasDefinedSourceLocale ? i18nOptions.sourceLocale : undefined,
     );
 
-    errors = result.errors;
-    warnings = result.warnings;
+    executionResult.addErrors(result.errors);
+    executionResult.addWarnings(result.warnings);
     prerenderedRoutes = result.prerenderedRoutes;
     executionResult.outputFiles.push(...result.additionalOutputFiles);
     executionResult.assetFiles.push(...result.additionalAssets);
   }
-
-  await printWarningsAndErrorsToConsoleAndAddToResult(context, executionResult, warnings, errors);
 
   if (prerenderOptions) {
     executionResult.addOutputFile(
@@ -307,6 +295,8 @@ export async function executeBuild(
     estimatedTransferSizes,
   );
 
+  await logMessages(context, executionResult);
+
   // Write metafile if stats option is enabled
   if (options.stats) {
     executionResult.addOutputFile(
@@ -317,21 +307,4 @@ export async function executeBuild(
   }
 
   return executionResult;
-}
-
-async function printWarningsAndErrorsToConsoleAndAddToResult(
-  context: BuilderContext,
-  executionResult: ExecutionResult,
-  warnings: string[],
-  errors: string[],
-): Promise<void> {
-  const errorMessages = errors.map((text) => ({ text, location: null }));
-  if (errorMessages.length) {
-    executionResult.addErrors(errorMessages);
-  }
-
-  await logMessages(context, {
-    errors: errorMessages,
-    warnings: warnings.map((text) => ({ text, location: null })),
-  });
 }

--- a/packages/angular_devkit/build_angular/src/tools/esbuild/bundler-execution-result.ts
+++ b/packages/angular_devkit/build_angular/src/tools/esbuild/bundler-execution-result.ts
@@ -38,6 +38,7 @@ export class ExecutionResult {
   outputFiles: BuildOutputFile[] = [];
   assetFiles: BuildOutputAsset[] = [];
   errors: (Message | PartialMessage)[] = [];
+  warnings: (Message | PartialMessage)[] = [];
   externalMetadata?: ExternalResultMetadata;
 
   constructor(
@@ -53,8 +54,32 @@ export class ExecutionResult {
     this.assetFiles.push(...assets);
   }
 
-  addErrors(errors: (Message | PartialMessage)[]): void {
-    this.errors.push(...errors);
+  addError(error: PartialMessage | string): void {
+    if (typeof error === 'string') {
+      this.errors.push({ text: error, location: null });
+    } else {
+      this.errors.push(error);
+    }
+  }
+
+  addErrors(errors: (PartialMessage | string)[]): void {
+    for (const error of errors) {
+      this.addError(error);
+    }
+  }
+
+  addWarning(error: PartialMessage | string): void {
+    if (typeof error === 'string') {
+      this.warnings.push({ text: error, location: null });
+    } else {
+      this.warnings.push(error);
+    }
+  }
+
+  addWarnings(errors: (PartialMessage | string)[]): void {
+    for (const error of errors) {
+      this.addWarning(error);
+    }
   }
 
   /**

